File: chapters/07_compaction_and_pruning/v1_hard_truncate.ts
// TypeScript mirror for Chapter 07: Compaction and Pruning (v1)
// Run with: npx ts-node v1_hard_truncate.ts

function hardTruncate(messages: string[], keepLast: number = 4): string[] {
  if (keepLast <= 0) {
    return [];
  }
  return messages.slice(-keepLast);
}

function main(): void {
  const data = Array.from({ length: 7 }, (_, i) => `m${i + 1}`);
  console.log(hardTruncate(data, 3));
}

main();

File: chapters/07_compaction_and_pruning/v1_hard_truncate.py
from __future__ import annotations


def hard_truncate(messages: list[str], keep_last: int = 4) -> list[str]:
    if keep_last <= 0:
        return []
    return messages[-keep_last:]


def main() -> None:
    data = [f"m{i}" for i in range(1, 8)]
    print(hard_truncate(data, keep_last=3))


if __name__ == "__main__":
    main()
